### DISCUSSION TITLE
Local folder not under git and I would like to link it to an existing github project

### DISCUSSION THREAD

**Discussion by AndrewHobson** [community/community#54617] [New to GitHub] [answered]
Hello,

Here is my situation.
I once installed a website project directly on my Ubuntu server by coping the files I downloaded from the people who created this project. At that time it was not on github.

I used it for years and today, I need to catch up and install some updates... And this project is now hosted on github.

What I would like to do is like initialise my local folder with git, then compare it with the project on github and then get my local folder in sync with a specific branch on github.

Is this possible ?

Where I am afraid is : I don't want to mess up the project on github. I want only the newer files to get downloaded in one direction. From github => to my local folder.

May I please ask someone for some help on this task ?

Thanks a lot.

Andrew

**Comment by airtower-luna** [comment, marked as answer]
> What I would like to do is like initialise my local folder with git, then compare it with the project on github and then get my local folder in sync with a specific branch on github.

That is possible, you just have to be careful not to overwrite your local files in the process (at least not without making sure it's what you want first). Making a backup first is always a good idea. :warning: 

The process will be:

* Create a new Git repository in your local directory (`git init`)
* Add the repository on GitHub as a *remote* (`git remote add [...]`)
* `git fetch` from the remote. After that your local Git repository has the data available in the remote one (but not checked out!).
* I recommend renaming the initially created branch to the same name as the upstream branch you want to use (if necessary), to avoid confusion (`git branch -m NAME`)
* Now you can *soft* reset to the remote branch you want to use (`git reset --soft REMOTE/NAME`). A soft reset makes your current branch point at the commit you provide (in this case the HEAD of the remote branch) *without touching your working tree files*.
* After that, `git diff` will show you the difference between your files and the latest commit on the current branch as usual. :slightly_smiling_face: 

> I don't want to mess up the project on github.

It sounds like other people run that project. As long as you don't have write access to the repository on GitHub that can't happen, any attempt to push your changes would simply be rejected. :wink:

**Reply by AndrewHobson** [reply]
Hi,

I need your good explanations again !
Well, I did this :  git remote add Name https://github.com/blablabla/Name.git

But where or how do I see that I prepared my local git to get this remote repository ?
I looked everywhere but I don't see no reference to Name...

But if I do again the same command it says fatal error the distant Name already exist.

I need to understand before going to the next stage.
And the git branch -r command doesn't show nothing.

Thanks

**Reply by airtower-luna** [reply]
*Name* in that command is the name of the *remote*. Use `git remote -v` to get a good overview of existing remotes, or `git remote show NAME` to see a specific one with more detail. `git branch -r` won't show anything until you *fetch* the remote.

See also: https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes

**Reply by AndrewHobson** [reply]
Ok, I understand a little more.
I did my : git fetch NAME, and it went well (but I also did something stupid that I will detail later...)
And I experimented the : git remote -v / git remote show NAME / git branch -r
All worked well.
But in your initial reply you suggested to me : I recommend renaming the initially created branch to the same name as the upstream branch you want to use (if necessary), to avoid confusion (git branch -m NAME)
Does thit mean that I need to rename my git working directory (is this what you intend in my "initially created branch"?) to the same name as the base branch I got the files from. If I undesrtand correctly, can you help me with this git branch -m NAME command ? And where do I see the name of my initially created branch ?
Also I tried simply : git diff, and I got no result. I guess I must use some arguments?
Also, here is the stupid thing I've done: I did :
git fetch https://github.com/blablabla/NAME.git
instead of simple : git fetch NAME
So what did the first command do exactly to my local repository ?
I know, I'm asking a lot...

**Reply by airtower-luna** [reply]
> But in your initial reply you suggested to me : I recommend renaming the initially created branch to the same name as the upstream branch you want to use (if necessary), to avoid confusion (git branch -m NAME)
> Does thit mean that I need to rename my git working directory (is this what you intend in my "initially created branch"?) to the same name as the base branch I got the files from. If I undesrtand correctly, can you help me with this git branch -m NAME command ? And where do I see the name of my initially created branch ?

I'm talking about the Git branch there, that has absolutely nothing to do with the name of the directory containing your repository (Git doesn't care about that). When you call `git init` it creates an initial branch in the new repository. After fetching you'll want to `git reset --soft` that branch to be equal to the remote branch you want to compare your local data with. My point is that it'd make sense to name your local branch the same as the remote one (e.g. if you want to track `origin/dev` it'd be confusing if your local branch is named `main` rather than `dev`, though if you want it that way it'll work). `git branch -m NEW_NAME` just renames the current branch.

You might also want to do `git branch --set-upstream-to=remote/branch` to make your local branch properly track the remote branch, that way `git pull` will work as usual. You don't have to though, you can still work with `git merge` manually if you want to bring in updates in the future.

> Also I tried simply : git diff, and I got no result. I guess I must use some arguments?

I assume you haven't done the soft reset yet, so you still have the (empty) initial branch.

> Also, here is the stupid thing I've done: I did :
> git fetch https://github.com/blablabla/NAME.git

That doesn't hurt, assuming the URL was the same as for the remote you've added at most you've pre-fetched the data. The most common use of fetching a URL is if you don't work with the other repository regularly and just fetch once-off, see the third [example in the `git fetch` documentation](https://git-scm.com/docs/git-fetch#_examples).

**Reply by AndrewHobson** [reply]
Hello,

Ok, so I tried out the command git branch -m NEW_NAME, but for the moment I reverted back to let my initial branch named master.

Sorry but I still don't understand, please stay calm and patient with me ;-)  And maybe one day I will stop bothering you !
I did the command: git fetch NAME
It worked well because when I do : git branch -r   I get a result
And tried : git reset --soft   but doesn't seem to have done something. I tried specifing the remote branch and name. But I'm not sure if I'im using it correctly... Can you pleas explain again ? And I don't get the concept.
Now I have what I call my "master" => The initial branch it created after doing the git init. And I have added the distant branches with the git remote add AND git fetch. So now this command git reset --soft  will sort of prepare my "master" to act like it was at the same stage as my remote branch but without changing my files, just letting me being able to do a git diff to see the differences and letting me decide if I want to apply or not ? Am I undesrtandign correctly ? SO when I use the command git reset --soft  It's obvious it's working on my initial branch "master" but I must precise the remote branch I want to reset to ? No ?
 
Also the : git diff   I understand it will compare my initial branch (created by git init) with the remote branch I will give as argument. Is my understanding correct ?

Thanks a lot.

**Reply by airtower-luna** [reply]
> So now this command git reset --soft will sort of prepare my "master" to act like it was at the same stage as my remote branch but without changing my files, just letting me being able to do a git diff to see the differences and letting me decide if I want to apply or not ?

That's the plan, yes. Apparently I made a mistake there, I just tried the same thing (`git reset --soft origin/main` after creating a new, empty repository and fetching one of my repositories into it). All my changes (well, deleting everything, because the empty repo had no files) were already marked as staged, it took another `git reset` (without the `--soft`) to undo that. So that'll probably work for you, too. :slightly_smiling_face: 

> SO when I use the command git reset --soft It's obvious it's working on my initial branch "master" but I must precise the remote branch I want to reset to ? No ?

Yes, you need to provide the branch, tag, or commit that you want to reset the current branch to. In your case that'll be the remote branch (e.g. `origin/main`). You can use `git log` to see if it worked, if yes it should show the same history as the remote branch.

> Also the : git diff I understand it will compare my initial branch (created by git init) with the remote branch I will give as argument. Is my understanding correct ?

By default (without any other parameters) `git diff` compares your working tree with your *index* (current commit plus any staged changes). You can tell it to compare other things on the command line, e.g. `main..origin/main` will compare the HEAD commit your local `main` branch with the remote `origin/main` branch.

**Reply by AndrewHobson** [reply]
It worked for me too ;-)

Thanks !

Some more questions... No retirement for you yet ;-)
I see there are some diffs. What must I do to actually accept them and go to this branch ?
I was on version 3.4.0a and the last stage of this branch 3.4.0 is the version 3.4.0d and I am Ok to move forward.

Another thing, afterwards, the concept is the same for me. I can git reset --soft to the next branch, git diff, and if I'm ok do the commands your are going to explain to me at my above question ?

Thanks for the top class help.

**Reply by airtower-luna** [reply]
> I see there are some diffs. What must I do to actually accept them and go to this branch ?

Now you're at the point where you can use the same commands as you would with any other repository, e.g. `git checkout` a specific file to set it to the version in the current commit. Or a hard reset if you want to discard *all* your changes and use only the remote version.

> Another thing, afterwards, the concept is the same for me. I can git reset --soft to the next branch, git diff, and if I'm ok do the commands your are going to explain to me at my above question ?

Technically you could, but again it's probably better to use the usual Git workflow, and pull/merge the updated remote branch.

What's going to be more interesting is if you want to keep *some* of your changes. In that case you might want to start your own branch, and either merge or rebase on top of remote changes.

**Reply by AndrewHobson** [reply]
I was able to do what I wanted.
You were really a tremendous kick-start for me. 
So thankfull. Sure ready to learn more and experiment.
If you come around Switzerland one time, I owe you a beer ;-) Or a few beers even !

**Reply by airtower-luna** [reply]
Great, will keep that in mind! :grin: 

Please don't forget to mark the discussion as answered, lots of answers to choose from. :wink:

**Comment by AndrewHobson** [comment]
Thanks for your reply.

Just one thing I need confirmation, when you say : 
Add the repository on GitHub as a remote (git remote add [...])

Does this mean that I need to do a : git remote add => to the official repository of the development team of this specific project ?
So this makes a sort of link between my local git and the official repo on github ?
Is my understanding correct ?

Thanks

**Comment by airtower-luna** [comment]
> Does this mean that I need to do a : git remote add => to the official repository of the development team of this specific project ?

Yes (assuming that's where you want to get the current version from).

> So this makes a sort of link between my local git and the official repo on github ?

Sort of. It tells your *local* Git repository that there's a *remote* repository (hence the name) holding the same project at that location. After that you can easily fetch the history of the repository that the remote has. You can have as many remotes as you want (that's why you have to give `git remote add` a name, too, so you can identify each remote), and if you have write access you can also push your changes to a remote.

See also: [Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)

**Comment by AndrewHobson** [comment]
Thanks for taking the time to explain. I appriciate ;-)

I will eperiment all this. After making a backup of course !

**Reply by NLPsixteen** [reply]
This is what i have looked after here, i am 56 years old and has have gaming/different focus within pc, even if you really not can call ZX81 as a pc :) but it was my first edjucation , but i was 14 and all found it super cool, dont sure if it was the first "pc" some say it but whatever whatever, i will write a little more about digital history though and then, was stressed and also little afraid here in beginning and still are, have only been here for like 3 days,am new to this all new codes and more,i had an accident 2015 kost my right arm and lot of more damages,was in Hospital 13 month, then all year until now fight against all medication and still have the psychological to fight for by learn about them, like bipolär,ptsd,add, depression and more so i felt also me afraid here when i saw how far back i was though i been away like 15 years from handle pc in legal purpose, i have already done a couple of wrong things here,and darken yelled at me, lol, no he informed me, i and many of my friends know about him,thik i have had sd maid sins 10 tears back and tryed out other apps from him, so i hope that he is understandable that i will be slower than other, i type so slow now that i get Angry, i would like to go an edjucation to learn pythonn nowday s it seems like all know how to code python, but i hope darken let me do wrong, especially in beginning but i want to learn the legal and serious way about codes, maybe because i got old but i starred gaming when i was about 8-10 and never got bored of all different technical things, have had 2 computer company, still one left but sleeping beaquorse 2 fast law changes in Sweden about streaming.
Thanks to you who started this community and have fun all.

**Comment by Katarzynamm** [comment]
Chciałabym płacić kartą Coinbase ale nie umiem na nią wpłacić np USD? Kto pomoże??
